aptos-move/aptos-e2e-comparison-testing/src/execution.rs

```diff
@@ -202,152 +233,180 @@ impl Execution {
                 &txn_idx,
                 compiled_package_cache,
                 compiled_package_cache_v2,
+                None,
             );
         }
         Ok(())
     }
 
-    fn execute_and_compare(
+    pub(crate) fn execute_and_compare(
         &self,
         cur_version: Version,
-        state: &FakeDataStore,
-        features: &Features,
+        state: FakeDataStore,
         txn_idx: &TxnIndex,
         compiled_package_cache: &HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
         compiled_package_cache_v2: &HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
+        debugger: Option<Arc<dyn AptosValidatorInterface + Send>>,
     ) {
         let mut package_cache_main = compiled_package_cache;
         let package_cache_other = compiled_package_cache_v2;
         if self.execution_mode.is_v2() {
             package_cache_main = compiled_package_cache_v2;
         }
         let res_main_opt = self.execute_code(
-            state,
-            features,
+            cur_version,
+            state.clone(),
             &txn_idx.package_info,
             &txn_idx.txn,
             package_cache_main,
+            debugger.clone(),
         );
         if self.execution_mode.is_compare() {
             let res_other_opt = self.execute_code(
+                cur_version,
                 state,
-                features,
                 &txn_idx.package_info,
                 &txn_idx.txn,
                 package_cache_other,
+                debugger.clone(),
             );
-            Self::print_mismatches(cur_version, &res_main_opt.unwrap(), &res_other_opt.unwrap());
+            self.print_mismatches(cur_version, &res_main_opt.unwrap(), &res_other_opt.unwrap());
         } else {
             let res = res_main_opt.unwrap();
             if let Ok(res_ok) = res {
-                println!(
+                self.output_result_str(format!(
                     "version:{}\nwrite set:{:?}\n events:{:?}\n",
                     cur_version, res_ok.0, res_ok.1
-                );
+                ));
             } else {
-                println!(
+                self.output_result_str(format!(
                     "execution error {} at version: {}, error",
                     res.unwrap_err(),
                     cur_version
-                );
+                ));
             }
         }
     }
 
     fn execute_code(
         &self,
-        state: &FakeDataStore,
-        features: &Features,
+        version: Version,
+        state: FakeDataStore,
         package_info: &PackageInfo,
         txn: &Transaction,
         compiled_package_cache: &HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
+        debugger_opt: Option<Arc<dyn AptosValidatorInterface + Send>>,
     ) -> Option<Result<(WriteSet, Vec<ContractEvent>), VMStatus>> {
         let executor = FakeExecutor::no_genesis();
         let mut executor = executor.set_not_parallel();
-        *executor.data_store_mut() = state.clone();
```

Comment:
I don't think we need a FakeExecutor her, why not just do `AptosVM::execute_block_no_limit(&vec![txn], &state)`? Then you a void coupling the execution with data state, and can use whatever `state` you want, either using V1 code or V2 code

Reply:
Plus our testing is a mess, I would trust `execute_block_no_limit` API more than `exec_...` variants

Reply:
The benefit of this API is that it 1) doesn't commit to state, so you can reuse the same state version for multiple executions, and 2) all things like aggregators are handled

Reply:
Thanks for the advice, @georgemitenkov. One concern I have is that for early txns, VM_BINARY_FORMAT_V6 is not enabled but the compiler only generates V6 bytecode. Is there a way to solve this issue?